onnxruntime/core/framework/session_state_utils.cc

```diff
@@ -339,10 +267,17 @@ common::Status SaveInitializedTensors(
     const auto entry = initialized_tensors_to_allocate.find(ort_value_index);
     ORT_ENFORCE(entry != initialized_tensors_to_allocate.end(),
                 "OrtValue index: ", ort_value_index, " from initializer_allocation_order not found among initialized tensors");
-    if (!(utils::HasExternalData(*entry->second) && exec_plan.GetLocation(ort_value_index).Type() == OrtDevice::CPU)) {
-      // can not trace string tensor
-      ORT_ENFORCE(entry->second->data_type() != ONNX_NAMESPACE::TensorProto_DataType_STRING, "Can not trace string tensor");
-      ORT_RETURN_IF_ERROR(planner.Trace(entry->first, entry->second));
+    const auto* tensor_proto = entry->second;
+
+    // We need to check if this is an external data on disk
+    // in memory we can still trace it
+    const bool trace_allocation = (exec_plan.GetLocation(ort_value_index).Type() != OrtDevice::CPU) ||
+                                  (!utils::HasExternalData(*tensor_proto) || utils::HasExternalDataInMemory(*tensor_proto));
```

Comment:
Do we want to trace an in-memory initializer if we didn't copy it? There's no allocation in that case is there?